### PR TITLE
[#ICC-80]  Add adapter from thirdparty message to pn messsage

### PR DESCRIPTION
**Pull request by fabriziopapi**
Add a node fetch implementations that will adapt a call to a third-party API to a Piattaforma Notifiche one. This adapter will allow us to use the generic third-party message implementations with the specific PN api, without further changes on PN api.

#### List of Changes

add pnFetch adapter
add pathParams decoder

#### Motivation and Context

Call PN api using the third-party definitions.

#### How Has This Been Tested?

unit tests

#### Types of changes

- [ ] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/adapters/pnFetch.ts [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+);
+
+const WithFiscalCode = t.interface({ fiscal_code: NonEmptyString });
+type WithFiscalCode = t.TypeOf<typeof WithFiscalCode>;
+
+const retrieveNotificationDetails = (
+  origFetch: typeof fetch,
+  pnUrl: string,
+  pnApiKey: string,
+  headers: WithFiscalCode,
```
Comment (AleDore) [inline src/adapters/pnFetch.ts:50]: why not passing directly `fiscalCode: FiscalCode` here?

File: src/adapters/pnFetch.ts [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+    TE.mapLeft(errorResponse),
+    TE.toUnion
+  )();
+
+const getPnDocumentUrl = (
+  origFetch: typeof fetch,
+  pnUrl: string,
+  pnApiKey: string,
+  url: string,
+  headers: WithFiscalCode
```
Comment (AleDore) [inline src/adapters/pnFetch.ts:173]: ditto

File: src/adapters/pnFetch.ts [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+              "x-pagopa-cx-taxid": headers.fiscal_code
+            }),
+          E.toError
+        ),
+        TE.chainEitherK(E.mapLeft(errorsToError)),
+        TE.chain(
+          TE.fromPredicate(
+            r => r.status === 200,
+            r =>
+              Error(`Failed to fetch PN SentNotificationDocument: ${r.status}`)
```
[the remark this answers is not in the document]
Comment (fabriziopapi) [reply src/adapters/pnFetch.ts:197]: We have to define a standard way to track the errors from proxied call: if we simply re-send the error code we do not known if the code is coming from io-backend or from pn (eg. if PN sent us a 401, we will forward it to the client without telling him that the auth problem is on pn and not in io-backend).

File: src/adapters/pnFetch.ts (hunk 1 above)
Comment (fabriziopapi) [reply src/adapters/pnFetch.ts:50]: done. thank you!

File: src/adapters/__tests__/pnFetch.test.ts [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+import { RecipientTypeEnum } from "../../../generated/piattaforma-notifiche/NotificationRecipient";
+import { NotificationFeePolicyEnum, PhysicalCommunicationTypeEnum } from "../../../generated/piattaforma-notifiche/NewNotificationRequest";
+import { NotificationStatusEnum } from "../../../generated/piattaforma-notifiche/NotificationStatus";
+import {
+    Client as PnClient
+} from "../../../generated/piattaforma-notifiche/client";
+import { NotificationAttachmentDownloadMetadataResponse } from "../../../generated/piattaforma-notifiche/NotificationAttachmentDownloadMetadataResponse";
+import { pipe } from "fp-ts/lib/function";
+import { Response as NodeResponse } from "node-fetch";
+const util = require('util');
```
[the remark this answers is not in the document]
Comment (fabriziopapi) [reply src/adapters/__tests__/pnFetch.test.ts]: done, thank you!

**Pull request merged**